cli/module_registry.go

```diff
@@ -323,12 +264,62 @@ func (m *ModuleID) toString() string {
 	return fmt.Sprintf("%s:%s", m.Prefix, m.Name)
 }
 
+// resolveModuleIDFromMultipleSources tries to parse the manifestNameEntry to see if it is a valid moduleID with a prefix
+// if it is not, it uses the publicNamespaceArg and orgIDArg to determine what the moduleID prefix should be.
+func resolveModuleIDFromMultipleSources(
+	c *cli.Context,
+	client *AppClient,
+	manifestNameEntry,
+	publicNamespaceArg,
+	orgIDArg string,
+) (ModuleID, error) {
+	moduleID, err := parseModuleID(manifestNameEntry)
+	if err != nil {
+		return ModuleID{}, err
+	}
+	if moduleID.Prefix != "" {
+		if publicNamespaceArg != "" || orgIDArg != "" {
+			org, err := resolveOrg(client, publicNamespaceArg, orgIDArg)
+			if err != nil {
+				return ModuleID{}, err
+			}
+			expectedOrg, err := getOrgByModuleIDPrefix(client, moduleID.Prefix)
+			if err != nil {
+				return ModuleID{}, err
+			}
+			if org.GetId() != expectedOrg.GetId() {
+				// This is almost certainly a user mistake
+				// Preferring org name rather than orgid here because the manifest probably has it specified in terms of
+				// public_namespace so returning the ids would be frustrating
+				return ModuleID{}, errors.Errorf("The meta.json specifies a different org %q than the one provided via args %q",
+					org.GetName(), expectedOrg.GetName())
+			}
+			fmt.Fprintln(c.App.Writer, "The module's meta.json already specifies a full module id. Ignoring public-namespace and org-id arg")
+		}
+		return moduleID, nil
+	}
+	// moduleID.Prefix is empty. Need to use orgIDArg and publicNamespaceArg to figure out what it should be
+	org, err := resolveOrg(client, publicNamespaceArg, orgIDArg)
+	if err != nil {
+		return ModuleID{}, err
+	}
+	if org.PublicNamespace != "" {
+		moduleID.Prefix = org.PublicNamespace
+	} else {
+		moduleID.Prefix = org.Id
+	}
```

Comment:
Could just specify the `org.Id` but I suspect that this will help give better error msgs throughout the system

Reply:
I like using namespace if exists. I think we should continue to build that namespace takes precedence over orgId. (like doing a redirect in the frontend as well).